[Step2] 문자열 계산기

jang1231111 opened the pull request:
코틀린 문법과 테스트 코드 모두 익숙치 않아 , 시간도 꽤 걸리고 많이 미숙한 것 같습니다.
피드백 주시면 열심히 공부하도록 하겠습니다. 잘 부탁드립니다.

jang1231111 committed:
[Step2] String Calculator

YooInKeun commented on src/main/kotlin/Calculator.kt:
```diff
@@ -0,0 +1,73 @@
+    fun calculate(expression: String?): Int {
+        nullOrEmptyCheck(expression)
+
+        val arr_experssion: Array<String> =
```
코틀린에서 변수는 Camel Case로 작성해주는 것이 컨벤션인데요~ ex) `arrExpression`
전체적으로 컨벤션에 맞게 수정해보는 것은 어떨까요?

YooInKeun commented on src/main/kotlin/Calculator.kt:
```diff
@@ -0,0 +1,73 @@
+        return result
+    }
+
+    fun operatorCheck(operando1: Int, arr_fomula: Array<String>): Int {
```
코드 전체적으로 접근 제어자를 신경 써주시면 더욱 좋을 거 같아요~
```suggestion
    private fun operatorCheck(operando1: Int, arr_fomula: Array<String>): Int {
```

YooInKeun commented on src/main/kotlin/Calculator.kt:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    fun nullOrEmptyCheck(expression: String?) {
+        if (expression == null || (expression == "")) {
```
코틀린에서 제공하는 메소드를 활용하면 좀 더 직관적이고 간결하게 표현 가능할 거 같아요! 
ex) `expression.isNullOrBlank`

YooInKeun commented on src/main/kotlin/Calculator.kt:
```diff
@@ -0,0 +1,73 @@
+
+    fun nullOrEmptyCheck(expression: String?) {
+        if (expression == null || (expression == "")) {
+            throw IllegalArgumentException("계산식이 NULL 혹은 공백입니다.")
```
require() 문을 활용해서 좀 더 코틀린스럽게 예외 처리를 해보는 것도 좋을 거 같아요!
https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/require.html

YooInKeun commented on src/main/kotlin/javatest.java:
```diff
@@ -0,0 +1,50 @@
+public class javatest {
```
테스트용으로 작성해보신 코드는 제외시키는 게 좋을 거 같네요~

YooInKeun commented on src/main/kotlin/Calculator.kt:
```diff
@@ -0,0 +1,73 @@
+        var result = operando1
+
+        while (i < arr_fomula.size) {
+            if (finishOperandoCheck(i, arr_fomula.size - 1)) {
```
객체지향 생활체조 원칙의 첫 번째 원칙이 위반 됐는데요~
들여쓰기를 줄여보는 방향으로 수정하면 좋을 거 같습니다

```
객체지향 생활체조 원칙
1. 한 메서드에 오직 한 단계의 들여쓰기만 한다.
```

이때 들여쓰기를 줄이는 방법에는 메소드 분리나 클래스 분리가 있을 거 같은데요~
**사칙 연산을 하는 책임을 별도 클래스로 분리해서 작업**해보는 건 어떨까요?

Calculator 클래스는 수식(expression)을 받는 역할을 갖고
수식에 대한 실제 연산은 별도 클래스(ex. Operator)에서 진행하면
**책임이 분리되어 유지보수하기 좋고 테스트도 용이한 구조**가 될 거 같아서요~
책임에 따라 클래스를 하나씩 분리해보면서 차근차근 진행하시는 걸 추천드립니다

해당 부분 진행이 어려우시면 편하게 DM 주셔도 좋고, 다른 분들의 코드를 참고해보셔도 좋습니다!

YooInKeun commented on src/test/kotlin/CalculatorTest.kt:
```diff
@@ -0,0 +1,59 @@
+import org.junit.jupiter.api.assertThrows
+import kotlin.IllegalArgumentException
+
+class CalculatorTest {
```
테스트 코드를 given(준비), when(실행), then(검증) 패턴에 맞게 구분지어서 작성해주는 것이 좀 더 읽기 쉬울 거 같아요
줄바꿈이나 주석으로 표시해주면 좋을 거 같습니다~

https://brunch.co.kr/@springboot/292

YooInKeun commented on src/main/kotlin/Calculator.kt:
```diff
@@ -0,0 +1,73 @@
+            expression?.split(" ".toRegex())?.dropLastWhile { it.isEmpty() }?.toTypedArray() as Array<String>
+        val operando1 = arr_experssion[0].toInt()
+
+        val result = operatorCheck(operando1, arr_experssion)
```
마이너하지만 early return 해서 불필요한 코드를 줄일 수 있을 거 같아요~
```suggestion
        return operatorCheck(operando1, arr_experssion)
```

YooInKeun commented on src/main/kotlin/Calculator.kt:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    fun divide(operando1: Int, operando2: Int): Int {
+        return operando1 / operando2
```
0으로 나눌 때의 예외 처리도 해주면 좋을 거 같네요 ㅎㅎ

YooInKeun commented on src/test/kotlin/CalculatorTest.kt:
```diff
@@ -0,0 +1,59 @@
+class CalculatorTest {
+
+    @Test
+    fun `2 더하기 3은 5이다`() {
```
사칙연산에 대한 테스트를 `@ParameterizedTest`를 통해 한 메소드에 작성하면 한 곳에서 확인하기 좋을 거 같아요 ㅎㅎ

ex)
```kotlin
@ParameterizedTest
@CsvSource(
    "1 + 2 = 3",
    "2 + 3 * 4 / 2 = 10",
    "4 - 2 / 2 * 5 + 1 = 6",
    delimiter = '='
)
fun `계산기 정상 동작`(expression: String, result: String) {
    // when, then
    assertThat(Calculator().calculate(expression)).isEqualTo(result.toInt())
}
```

YooInKeun reviewed (changes requested):
2단계 미션 진행하시느라 고생 많으셨습니다 👍 👍 
몇가지 코멘트 남겼으니 확인해보시고 재요청 주세요~ 🙇‍♂️ 
화이팅입니다~!!!

jang1231111 committed:
[step2] 피드백 수정사항

1. 변수 명 CamelCase에 맞게 수정
2. 메소드 접근제어자 명시
3. 사용하지않는 테스트 코드 제거
4. require 문 사용 , Null or Empty일 경우 IllegalArgumentException
5. 객체지향 생활체조 1 원칙 : 한 메서드에 오직 한 단계의 들여쓰기만 한다.
메서드나 클래스를 나누는게 익숙치 않아 , 이런식으로 하는것이 맞는지 모르겠네요.
메서드의 크기는 최대한 들여쓰기를 줄이기 위해 나눠봤고 ,
피드백 주신 대로
Operator클래스를 새롭게 만들어 본격적인 계산 연산이 이루어지는 클래스로 만들었고
Calculator 클래스에서는 수식의 문법등의 Exception을 체크하도록 만들어 보았습니다.
6. 테스트코드 given , when , thne 패턴에 구분짓게 수정
7. @parameterizedTest에 사칙연산에 대한 수식 배열을 만들어 한번에 정리

이런식으로 메세지를 남기면 되는지 모르겠으나 , 리뷰어님께서 주신 피드백에 맞춰 다시 코드를 작성해 보았습니다
틀린 부분이 있다면 다시 피드백 주시면 감사하겠습니다 !

jang1231111 committed:
require문 사용 추가

이전 커밋에서 추가된 내용이 있어 다시 커밋합니다 !

YooInKeun commented on src/main/kotlin/Calculator.kt:
```diff
@@ -0,0 +1,35 @@
+        return operator.operate(arrExpression)
+    }
+
+    private fun operatorCheck(operator: String, operand2: String) {
```
operator 예외 처리는 `Operator` 클래스에 있는 게 역할에 맞지 않을까요~?

YooInKeun commented on src/main/kotlin/Calculator.kt:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    private fun nullOrEmptyCheck(expression: String?) {
+        require(!expression.isNullOrEmpty()) { "계산식이 NULL 혹은 공백입니다." }
```
예외 처리 구문을 간결하게 수정해주셨군요 👍

YooInKeun commented on src/main/kotlin/Operator.kt:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    /** 연산 기호를 확인하고 그에 맞는 함수 호출 */
+    private fun operatorCheck(operand1: Int, operator: String, operand2: Int): Int {
```
메소드명으로 봤을 때는 무언가를 검사하는 로직 같은데, 실제 동작은 사칙연산을 하고 있네요~

좀 더 정확한 네이밍으로 수정해줘도 좋을 거 같습니다!

YooInKeun commented on src/main/kotlin/Operator.kt:
```diff
@@ -0,0 +1,61 @@
+    /** 연산 기호를 확인하고 그에 맞는 함수 호출 */
+    private fun operatorCheck(operand1: Int, operator: String, operand2: Int): Int {
+        var result = 0
+        when (operator) {
```
사칙연산을 연산 기호에 따라 깔끔하게 구현해주셨네요 👍 
이때 다음 단계로 다형성을 활용해서 사칙 연산을 구현해보는 것은 어떨까요~? 

문제 해결 그 자체도 중요하지만, 설계를 신경써주는 것이 유지보수와 재사용성 면에서 더욱 중요하다고 생각합니다

연산 기호와 그에 따른 실제 연산을 프로퍼티로 갖는 클래스를 만들어서 시간 되실 때 연습해보시는 것을 추천드립니다!!

이에 따른 예시 코드를 첨부드릴테니 혹시 이해가 안가시거나 도움이 필요하면 DM 주셔도 좋습니다!

```kotlin
enum class Operator(
    private val symbol: Char,
    val operator: (Int, Int) -> Int
) {

    PLUS('+', { num1, num2 -> num1 + num2 }),
    MINUS('-', { num1, num2 -> num1 - num2 }),
    MULTIPLICATION('*', { num1, num2 -> num1 * num2 }),
    DIVISION('/', { num1, num2 ->
        require(num2 != 0) { "나누는 수가 0이 될 수 없습니다." }
        num1 / num2
    });

    companion object {
        fun findOperator(symbol: Char): Operator {
            return Operator.values()
                .find { it.symbol == symbol }
                ?: throw IllegalArgumentException("존재하지 않는 수식 기호입니다. (symbol: $symbol)")
        }
    }
}
```

YooInKeun reviewed (approved):
2단계 피드백 반영하느라 고생많으셨습니다 👍 👍
추가 작업하신 부분에 대해 코멘트 남겼으니 확인 부탁드리고
3단계에 함께 반영해주시면 됩니다! 
본격적인 자동차 미션도 화이팅입니다 ~ 😄

YooInKeun commented:
추가적으로 미션 진행전에 요구사항을 README.md 에 작성해보시고 기능이 완료될 때마다 체크해보는 것을 추천드립니다
사전에 어떤 기능을 구현할지 정의하는 것이 설계에 좋은 밑거름이 되고 단계적으로 미션을 진행하시는데 도움이 될 거 같아요~

ex)

## 기능 요구사항

- [x] 입력 순서에 따라 계산하는 사칙 연산 계산기를 구현한다.
- [ ] ...

YooInKeun commented:
또한 커밋 하실 때, 커밋 메세지를 컨벤션에 맞춰서 작성해보시면 좋을 거 같아요~
https://gist.github.com/stephenparish/9941e89d80e2bc58a153
